make/Images.gmk

```diff
@@ -98,16 +111,48 @@ $(eval $(call SetupExecute, jlink_jdk, \
     WARN := Creating jdk image, \
     DEPS := $(JDK_JMODS) $(BASE_RELEASE_FILE) \
         $(call DependOnVariable, JDK_MODULES_LIST, $(JDK_IMAGE_SUPPORT_DIR)/_jlink_jdk.vardeps), \
-    OUTPUT_DIR := $(JDK_IMAGE_DIR), \
+    OUTPUT_DIR := $(JDK_LINK_OUTPUT_DIR), \
     SUPPORT_DIR := $(JDK_IMAGE_SUPPORT_DIR), \
-    PRE_COMMAND := $(RM) -r $(JDK_IMAGE_DIR), \
+    PRE_COMMAND := $(RM) -r $(JDK_LINK_OUTPUT_DIR), \
     COMMAND := $(JLINK_TOOL) --add-modules $(JDK_MODULES_LIST) \
-        $(JLINK_JDK_EXTRA_OPTS) --output $(JDK_IMAGE_DIR) \
+        $(JLINK_JDK_EXTRA_OPTS) --output $(JDK_LINK_OUTPUT_DIR) \
         $(JLINK_DISABLE_WARNINGS), \
 ))
 
 JLINK_JDK_TARGETS := $(jlink_jdk)
 
+ifeq ($(JLINK_PRODUCE_RUNTIME_LINK_JDK), true)
+
+  JDK_RUN_TIME_IMAGE_SUPPORT_DIR := $(SUPPORT_OUTPUTDIR)/images/jlink-run-time
+  JDK_JMODS_DIFF := $(JDK_RUN_TIME_IMAGE_SUPPORT_DIR)/jimage_packaged_modules_diff.data
+  JLINK_RUNTIME_CREATE_EXTRA += --create-linkable-runtime=$(JDK_JMODS_DIFF)
+
+  $(eval $(call SetupExecute, generate_jimage_diff, \
+      WARN := Generating the jimage to packaged modules diff, \
+      DEPS := $(jlink_jdk), \
+      OUTPUT_FILE := $(JDK_JMODS_DIFF), \
+      SUPPORT_DIR := $(JDK_RUN_TIME_IMAGE_SUPPORT_DIR), \
+      COMMAND := $(TOOL_JIMAGE_DIFF_TO_JMODS) $(IMAGES_OUTPUTDIR)/jmods \
+          $(JDK_LINK_OUTPUT_DIR)/lib/modules $(JDK_JMODS_DIFF), \
+  ))
+
+  $(eval $(call SetupExecute, jlink_runtime_jdk, \
+      WARN := Creating jdk image for runtime linking, \
+      DEPS := $(generate_jimage_diff), \
+      OUTPUT_DIR := $(JDK_IMAGE_DIR), \
+      SUPPORT_DIR := $(JDK_RUN_TIME_IMAGE_SUPPORT_DIR), \
+      PRE_COMMAND := $(RM) -r $(JDK_IMAGE_DIR), \
```

Comment:
This looks scary! Why the rm? Are you sure you are not racing with some other rule that tries to read from the JDK_IMAGE_DIR?

Reply:
That was modelled similar to `jdk_jlink` target. It also does the removal. When building with `--enable-runtime-link-image`, the flow is:

```
1. Link the initial jdk image (current `images/jdk`). Output is $(RL_INTERMEDIATE_IMAGE_DIR)
2. Generate the diff to the packaged modules (target `generate_jimage_diff`)
3. Do the final link creating a runtime linkable jimage with `--create-linkabel-runtime` into `JDK_IMAGE_DIR`.
```

All three steps should have appropriate dependencies on each other. So I don't think there is a race. If you see one please let me know! Thanks.

Reply:
I don't see a race. The `rm` was there in the original code and is no scarier in the modified version. The jdk image is constructed by a combination of targets and recipes. The first one to run has to be jlink, then we overlay more stuff on top. If the makefile dependency resolution concludes that we need to rerun the jlink step, we clear out the directory completely to make sure we aren't leaving anything in there that's no longer valid. This is basically a precaution to guarantee a correct incremental build. It's not incurring a big build time penalty as jlink would have overwritten all files it would have created anyway. However, if a module was removed, or a file was removed from a module, the `rm` helps guarantee that we don't include such a removed file from a previous build attempt in the final image.

Reply:
Ok, good, thanks. I did not see the whole picture there.

Reply:
Marking this as resolved.